In [1]:
import urllib
import urllib.request
import datetime
import pandas as pd
import numpy as np
from pandas import read_csv  
import random as rd # generating random numbers
# from datetime import datetime
import os
# %matplotlib inline
# import matplotlib.pyplot as plt
import seaborn as sns

In [124]:
df = read_csv('nyc_mta_data.csv')

C:\Users\Shraddha Somani\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2717: DtypeWarning: Columns (9,10) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [125]:
df.describe()

C/A      UNIT       SCP         STATION  LINENAME  DIVISION  \
count   10267862  10267862  10267862        10267862  10267862  10267862   
unique       732       468       220             377       129         7   
top        PTH05      R549  00-00-00  34 ST-PENN STA         1       IRT   
freq       54629    131650   1028471          223582   1157633   3790711   

              DATE      TIME      DESC   ENTRIES  \
count     10267862  10267862  10267862  10267862   
unique         372     85384         3   6860803   
top     02/19/2016  04:00:00   REGULAR         0   
freq         34773    794799  10225225    134111   

        EXITS                                                                 
count                                            10267862                     
unique                                            6738931                     
top                                                     0                     
freq                                               258467

# Data Description

The data file consists of 10267862 rows and 11 columns namely:
1. C/A      = Control Area (A002)
2. UNIT     = Remote Unit for a station (R051)
3. SCP      = Subunit Channel Position represents an specific address for a device (02-00-00)
4. STATION  = Represents the station name the device is located at
5. LINENAME = Represents all train lines that can be boarded at this station. LINENAME 456NQR repersents train server for 4, 5, 6, N, Q, and R trains.
6. DIVISION = Represents the Line originally the station belonged to
    - IRT = Interborough Rapid Transit;
    - BMT = Brooklyn-Manhattan Transit;
    - IND = Independent Subway;
    - PTH = Port Authority Trans-Hudson (PATH);
    - RIT = Roosevelt Island Tram;
    - SRT = Staten Island Rapid Transit;
7. DATE     = Represents the date (MM-DD-YY)
8. TIME     = Represents the time (hh:mm:ss) for a scheduled audit event
9. DESc     = Represent the "REGULAR" scheduled audit event (Normally occurs every 4 hours)
10. ENTRIES  = The cumulative entry register value for a device
11. EXIST    = The cumulative exit register value for a device

In [126]:
df.head()

C/A  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST  NQR456W      BMT  12/24/2016  03:00:00   
1  A002  R051  02-00-00   59 ST  NQR456W      BMT  12/24/2016  07:00:00   
2  A002  R051  02-00-00   59 ST  NQR456W      BMT  12/24/2016  11:00:00   
3  A002  R051  02-00-00   59 ST  NQR456W      BMT  12/24/2016  15:00:00   
4  A002  R051  02-00-00   59 ST  NQR456W      BMT  12/24/2016  19:00:00   

      DESC  ENTRIES  \
0  REGULAR  5980509   
1  REGULAR  5980531   
2  REGULAR  5980613   
3  REGULAR  5980899   
4  REGULAR  5981382   

  EXITS                                                                 
0                                            2025806                    
1                                            2025818                    
2                                            2025912                    
3                                            2025965                    
4                                            2026001

In [127]:
df.tail()

C/A  UNIT       SCP        STATION LINENAME DIVISION        DATE  \
10267857  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R      RIT  01/01/2016   
10267858  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R      RIT  01/01/2016   
10267859  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R      RIT  01/01/2016   
10267860  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R      RIT  01/01/2016   
10267861  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R      RIT  01/01/2016   

              TIME     DESC ENTRIES  \
10267857  04:00:00  REGULAR    5554   
10267858  08:00:00  REGULAR    5554   
10267859  12:00:00  REGULAR    5554   
10267860  16:00:00  REGULAR    5554   
10267861  20:00:00  REGULAR    5554   

         EXITS                                                                 
10267857                                                230                    
10267858                                                230                    
10267859                                                230                    
10267860                                                230                    
10267861                                                230

# Data Modeling

In [2]:
from sklearn.model_selection import train_test_split

For data modeling, we split the data into - Train (70%) and Test (30%) 

In [22]:
df_train, df_test = train_test_split(df_final, test_size = 0.3)
# Dividing the df_train data again because the computer cannot handle huge amount of data
df_train1, df_test1 = train_test_split(df_train, test_size = 0.3)
df_train2, df_test2 = train_test_split(df_train1, test_size = 0.3)
df_train3, df_test3 = train_test_split(df_train2, test_size = 0.3)
df_train4, df_test4 = train_test_split(df_train3, test_size = 0.3)
df_train5, df_test5 = train_test_split(df_train4, test_size = 0.3)
df_train6, df_test6 = train_test_split(df_train5, test_size = 0.3)
df_train7, df_test7 = train_test_split(df_train6, test_size = 0.3)
df_train8, df_test8 = train_test_split(df_train7, test_size = 0.3)
df_train9, df_test9 = train_test_split(df_train8, test_size = 0.3)
df_train_final, df_test_final = train_test_split(df_train9, test_size = 0.3)

In [23]:
# Write the final train dataframe to csv file
df_train_final.to_csv('nyc_mta_train_final.csv', index = False)

In [24]:
# Write the final test dataframe to csv file
df_test_final.to_csv('nyc_mta_test_final.csv', index = False)

In [25]:
df_train = read_csv('nyc_mta_train_final.csv')

In [26]:
df_train.shape

(188974, 18)

In [27]:
df_test = read_csv('nyc_mta_test_final.csv')

In [28]:
df_test.shape

(80989, 18)

In [29]:
df_train.dtypes

C/A            object
UNIT           object
SCP            object
STATION        object
LINENAME       object
DIVISION       object
DATE           object
TIME           object
DESC           object
ENTRIES         int64
EXITS           int64
DATETIME       object
YEAR            int64
MONTH           int64
DAY             int64
WEEKDAY        object
HOUR            int64
TOTAL_COUNT     int64
dtype: object

### Feature Selection 

We perform feature selection to identify most relevant features (variables, predictors) for constructing our model. Feature selection techniques are used for four reasons:

- Simplification of models to make them easier to interpret by researchers/users
- Shorter training times
- To avoid the curse of dimensionality
- Enhanced generalization by reducing overfitting

Linear Regression & Principle Component Analysis (PCA) is performed to identify most significant variables that influences the entries and exits.

After performing the above two techniques in R, the factors affecting the Entries count are:

- C.A
- UNIT
- SCP
- STATION
- LINENAME
- DIVISION 
- MONTH
- HOUR

For predicting the Entries, Random Forest & XGBoost Linear Regressor is used.

### Random Forest

In [30]:
import sklearn
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn import preprocessing as pre
!pip install xgboost
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

In [31]:
df_model = df_train.loc[:,['C.A','UNIT','SCP','STATION','LINENAME','DIVISION','ENTRIES','MONTH','HOUR']]

In [32]:
df_model.dtypes

C.A         float64
UNIT         object
SCP          object
STATION      object
LINENAME     object
DIVISION     object
ENTRIES       int64
MONTH         int64
HOUR          int64
dtype: object

In [33]:
# Changing the data type for the model as the regressor accepts only integer values
number = LabelEncoder()

df_model['C.A'] = number.fit_transform(df_model['C.A'])
df_model['UNIT'] = number.fit_transform(df_model['UNIT'])
df_model['SCP'] = number.fit_transform(df_model['SCP'])
df_model['STATION'] = number.fit_transform(df_model['STATION'])
df_model['LINENAME'] = number.fit_transform(df_model['LINENAME'])
df_model['DIVISION'] = number.fit_transform(df_model['DIVISION'])

In [34]:
df_model.head()

C.A  UNIT  SCP  STATION  LINENAME  DIVISION    ENTRIES  MONTH  HOUR
0       0   295   45      256        80         1       3312      9     8
1  125975   271    0       57         4         2    5729268      8    14
2  125976   130   45       85        33         2       2695      5    21
3  125977    84   26      188        61         0    3621360     11    12
4  125978   243    2      266        95         0  135058151      6     8

In [35]:
# Train the data without ENTRIES
train_data = df_model.reset_index(drop = True)
y_labels = train_data['ENTRIES']
X_train_data = train_data.drop(['ENTRIES'], axis = 1)
print('Shape X :', X_train_data.shape)
print('Shape y :', y_labels.shape)

Shape X : (188974, 8)
Shape y : (188974,)


In [36]:
# Split the data into train & test
num_test = 0.20
X_train, X_validation, y_train, y_validation = train_test_split(X_train_data, y_labels, test_size=num_test, random_state=15)
print('X_train shape :', X_train.shape)
print('y_train shape :', y_train.shape)
print('X_validation shape :', X_validation.shape)
print('y_validation shape :', y_validation.shape)

X_train shape : (151179, 8)
y_train shape : (151179,)
X_validation shape : (37795, 8)
y_validation shape : (37795,)


In [37]:
# Choose the type of classifier. 
RFR = RandomForestRegressor()

# Choose some parameter combinations. We can also try different parameters to find the best model
parameters = {'n_estimators': [5, 10, 100],
              #'criterion': ['mse'],
              #'max_depth': [5, 10, 15], 
              #'min_samples_split': [2, 5, 10],
              'min_samples_leaf': [1,5]
             }

# Type of scoring used to compare parameter combinations
# We have to use RandomForestRegressor's own scorer (which is R^2 score)

# Run the grid search
grid_obj = GridSearchCV(RFR, parameters,
                        cv=5, #Determines the cross-validation splitting strategy /to specify the number of folds in a (Stratified)KFold
                        n_jobs=-1, #Number of jobs to run in parallel
                        verbose=1)
grid_obj = grid_obj.fit(X_train, y_train)

# Set the clf to the best combination of parameters
RFR = grid_obj.best_estimator_

# Fit the best algorithm to the data. 
RFR.fit(X_train, y_train)

Fitting 5 folds for each of 6 candidates, totalling 30 fits


[Parallel(n_jobs=-1)]: Done  30 out of  30 | elapsed:  7.6min finished


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_split=1e-07, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           n_estimators=100, n_jobs=1, oob_score=False, random_state=None,
           verbose=0, warm_start=False)

In [38]:
predictions = RFR.predict(X_validation)

#if we want to Re-scale, use this lines of code :
#predictions = predictions * (max_train - min_train) + min_train
#y_validation_RF = y_validation * (max_train - min_train) + min_train

#if not, keep this one:
y_validation_RF = y_validation

print('R2 score = ', r2_score(y_validation_RF, predictions), '/ 1.0')
print('MSE score = ', mean_squared_error(y_validation_RF, predictions), '/ 0.0')

R2 score =  0.976098655729 / 1.0
MSE score =  9.37405413393e+14 / 0.0
